### PR TITLE
DROOLS-1460 DMN extension for kie-server

### PR DESCRIPTION
**Pull request by tarilabs**
depends on droolsjbpm/droolsjbpm-build-bootstrap#402 .

### CODE REVIEW

**Comment by tarilabs** [comment]
```
19:41:45 [INFO] Reactor Summary:
19:41:45 [INFO] 
19:41:45 [INFO] Drools and jBPM integration multiproject ........... SUCCESS [  2.656 s]
19:41:45 [INFO] KIE Identity And Session Provider .................. SUCCESS [ 14.724 s]
19:41:45 [INFO] KIE :: Maven Plugin ................................ SUCCESS [04:37 min]
19:41:45 [INFO] Drools Benchmark ................................... SUCCESS [ 20.059 s]
19:41:45 [INFO] Drools RHQ JMX Plugin .............................. SUCCESS [ 14.589 s]
19:41:45 [INFO] Drools and jBPM examples ........................... SUCCESS [ 26.090 s]
19:41:45 [INFO] jBPM Simulation .................................... SUCCESS [ 52.959 s]
19:41:45 [INFO] Kie :: Spring ...................................... SUCCESS [03:25 min]
19:41:45 [INFO] Kie :: Aries Blueprint ............................. SUCCESS [ 48.548 s]
19:41:45 [INFO] KIE EAP :: Drools and jBPM Module layers for JBoss EAP SUCCESS [  0.147 s]
19:41:45 [INFO] KIE EAP :: Config .................................. SUCCESS [  7.920 s]
19:41:45 [INFO] KIE EAP :: Distribution ............................ SUCCESS [ 25.892 s]
19:41:45 [INFO] KIE EAP :: Installer ............................... SUCCESS [ 13.468 s]
19:41:45 [INFO] KIE EAP :: Integration Tests ....................... SUCCESS [ 31.066 s]
19:41:45 [INFO] KIE :: Infinispan .................................. SUCCESS [  0.204 s]
19:41:45 [INFO] KIE :: Infinispan :: Drools persistence ............ SUCCESS [01:05 min]
19:41:45 [INFO] KIE :: Infinispan :: jBPM persistence .............. SUCCESS [01:22 min]
19:41:45 [INFO] Drools :: JBoss Integration ........................ SUCCESS [ 10.043 s]
19:41:45 [INFO] KIE :: Execution Server ............................ SUCCESS [  0.165 s]
19:41:45 [INFO] KIE :: Execution Server :: Common .................. SUCCESS [ 16.819 s]
19:41:45 [INFO] KIE :: Execution Server :: API ..................... SUCCESS [ 38.442 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote .................. SUCCESS [  0.169 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: Client ........ SUCCESS [ 42.200 s]
19:41:45 [INFO] KIE :: Execution Server :: Controller .............. SUCCESS [  0.154 s]
19:41:45 [INFO] KIE :: Execution Server :: Controller :: API ....... SUCCESS [ 16.499 s]
19:41:45 [INFO] KIE :: Execution Server :: Services ................ SUCCESS [  0.202 s]
19:41:45 [INFO] KIE :: Execution Server :: Services :: Common ...... SUCCESS [ 30.197 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: JMS ........... SUCCESS [ 11.184 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: REST .......... SUCCESS [  0.072 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: REST :: Common  SUCCESS [ 12.801 s]
19:41:45 [INFO] KIE :: Execution Server :: Services :: jBPM Extension SUCCESS [ 28.839 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: REST :: jBPM .. SUCCESS [ 17.931 s]
19:41:45 [INFO] KIE :: Execution Server :: Services :: Drools Extension SUCCESS [ 11.452 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: REST :: Drools  SUCCESS [ 13.766 s]
19:41:45 [INFO] jBPM Process SVG Manipulation ...................... SUCCESS [ 14.721 s]
19:41:45 [INFO] KIE :: Execution Server :: Services :: jBPM UI Extension SUCCESS [ 19.058 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: REST :: jBPM UI SUCCESS [ 14.667 s]
19:41:45 [INFO] KIE :: Execution Server :: Services :: OptaPlanner Extension SUCCESS [ 11.837 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: REST :: OptaPlanner SUCCESS [ 14.221 s]
19:41:45 [INFO] KIE :: Execution Server :: Services :: Case Management Extension SUCCESS [ 16.330 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: REST :: Case Management SUCCESS [ 15.192 s]
19:41:45 [INFO] KIE :: Execution Server :: Services :: DMN Extension SUCCESS [ 13.252 s]
19:41:45 [INFO] KIE :: Execution Server :: Remote :: REST :: DMN ... SUCCESS [ 11.062 s]
19:41:45 [INFO] KIE :: Execution Server :: Router .................. SUCCESS [  0.182 s]
19:41:45 [INFO] KIE :: Execution Server :: Router :: Proxy ......... SUCCESS [ 19.891 s]
19:41:45 [INFO] KIE :: Execution Server :: Router :: Client ........ SUCCESS [ 26.457 s]
19:41:45 [INFO] KIE :: Execution Server :: Wars .................... SUCCESS [  0.075 s]
19:41:45 [INFO] KIE :: Execution Server :: Wars :: Distribution Wars SUCCESS [01:19 min]
19:41:45 [INFO] KIE :: Execution Server :: Wars :: Distribution .... SUCCESS [  6.080 s]
19:41:45 [INFO] KIE :: Execution Server :: Controller :: Implementation SUCCESS [ 14.622 s]
19:41:45 [INFO] KIE :: Execution Server :: Controller :: REST Implementation SUCCESS [ 10.995 s]
19:41:45 [INFO] KIE :: Execution Server :: Controller :: Test WAR .. SUCCESS [ 24.012 s]
19:41:45 [INFO] KIE :: Execution Server :: Tests ................... SUCCESS [  8.474 s]
19:41:45 [INFO] KIE :: Execution Server :: Tests :: Common Integration Tests SUCCESS [02:42 min]
19:41:45 [INFO] KIE :: Execution Server :: Tests :: Drools Integration Tests SUCCESS [03:01 min]
19:41:45 [INFO] KIE :: Execution Server :: Tests :: jBPM Integration Tests SUCCESS [11:44 min]
19:41:45 [INFO] KIE :: Execution Server :: Tests :: OptaPlanner Integration Tests FAILURE [02:14 min]
19:41:45 [INFO] KIE :: Execution Server :: Tests :: Drools and jBPM Integration Tests SKIPPED
19:41:45 [INFO] KIE :: Execution Server :: Tests :: Controller Integration Tests SKIPPED
19:41:45 [INFO] KIE :: Execution Server :: Tests :: Policies Integration Tests SKIPPED
19:41:45 [INFO] KIE :: Execution Server :: Tests :: Router Integration Tests SKIPPED
19:41:45 [INFO] KIE :: Execution Server :: Tests :: Case Id Generator Integration Tests SKIPPED
19:41:45 [INFO] KIE :: Execution Server :: Tests :: DMN Integration Tests SUCCESS [ 59.935 s]
19:41:45 [INFO] KIE :: OSGi Integration Parent Project ............. SUCCESS [  0.201 s]
19:41:45 [INFO] kie-karaf-itests-domain-model ...................... SUCCESS [ 13.212 s]
19:41:45 [INFO] KIE :: Karaf Integration Tests Kjar ................ SUCCESS [  9.366 s]
19:41:45 [INFO] KIE :: OSGi Integration ............................ SUCCESS [  8.942 s]
19:41:45 [INFO] KIE :: Karaf Features .............................. SUCCESS [ 10.737 s]
19:41:45 [INFO] KIE :: Karaf Integration Tests ..................... SUCCESS [02:56 min]
19:41:45 [INFO] KIE Performance Kit ................................ SUCCESS [ 16.975 s]
19:41:45 [INFO] Drools :: Android Integration ...................... SUCCESS [ 16.786 s]
19:41:45 [INFO] kie-security ....................................... SUCCESS [  8.017 s]
19:41:45 [INFO] Drools and jBPM integration distribution ........... SUCCESS [  6.825 s]
19:41:45 [INFO] BPMS distribution .................................. SUCCESS [ 17.591 s]
19:41:45 [INFO] BRMS distribution .................................. SUCCESS [  7.494 s]
19:41:45 [INFO] ------------------------------------------------------------------------
19:41:45 [INFO] BUILD FAILURE
19:41:45 [INFO] ------------------------------------------------------------------------
19:41:45 [INFO] Total time: 27:20 min (Wall Clock)
19:41:45 [INFO] Finished at: 2017-03-02T14:41:46-05:00
19:41:46 [INFO] Final Memory: 514M/1577M
19:41:46 [INFO] ------------------------------------------------------------------------
19:41:46 [ERROR] Failed to execute goal org.codehaus.cargo:cargo-maven2-plugin:1.6.2:stop (stop-container) on project kie-server-integ-tests-optaplanner: Execution stop-container of goal org.codehaus.cargo:cargo-maven2-plugin:1.6.2:stop failed: Failed to stop the WildFly 10.0.0.Final container. Failure when invoking CLI script, java returned 1 -> [Help 1]
19:41:46 org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.cargo:cargo-maven2-plugin:1.6.2:stop (stop-container) on project kie-server-integ-tests-optaplanner: Execution stop-container of goal org.codehaus.cargo:cargo-maven2-plugin:1.6.2:stop failed: Failed to stop the WildFly 10.0.0.Final container.
19:41:46 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:224)
19:41:46 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
19:41:46 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
19:41:46 	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
19:41:46 	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:188)
19:41:46 	at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call(MultiThreadedBuilder.java:184)
19:41:46 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
19:41:46 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
19:41:46 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
19:41:46 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
19:41:46 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
19:41:46 	at java.lang.Thread.run(Thread.java:745)
19:41:46 Caused by: org.apache.maven.plugin.PluginExecutionException: Execution stop-container of goal org.codehaus.cargo:cargo-maven2-plugin:1.6.2:stop failed: Failed to stop the WildFly 10.0.0.Final container.
19:41:46 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:143)
19:41:46 	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
19:41:46 	... 11 more
19:41:46 Caused by: org.codehaus.cargo.container.ContainerException: Failed to stop the WildFly 10.0.0.Final container.
19:41:46 	at org.codehaus.cargo.container.spi.AbstractLocalContainer.stop(AbstractLocalContainer.java:303)
19:41:46 	at org.codehaus.cargo.maven2.ContainerStopMojo.doExecute(ContainerStopMojo.java:47)
19:41:46 	at org.codehaus.cargo.maven2.AbstractCargoMojo.execute(AbstractCargoMojo.java:462)
19:41:46 	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
19:41:46 	... 12 more
19:41:46 Caused by: org.codehaus.cargo.container.ContainerException: Failure when invoking CLI script, java returned 1
19:41:46 	at org.codehaus.cargo.container.wildfly.internal.AbstractWildFlyInstalledLocalContainer.executeScriptFiles(AbstractWildFlyInstalledLocalContainer.java:245)
19:41:46 	at org.codehaus.cargo.container.wildfly.internal.AbstractWildFlyInstalledLocalContainer.executeScript(AbstractWildFlyInstalledLocalContainer.java:209)
19:41:46 	at org.codehaus.cargo.container.wildfly.internal.AbstractWildFlyInstalledLocalContainer.doStop(AbstractWildFlyInstalledLocalContainer.java:178)
19:41:46 	at org.codehaus.cargo.container.spi.AbstractInstalledLocalContainer.stopInternal(AbstractInstalledLocalContainer.java:305)
19:41:46 	at org.codehaus.cargo.container.spi.AbstractLocalContainer.stop(AbstractLocalContainer.java:282)
19:41:46 	... 15 more
```

Jenkins, test again

**Comment by tarilabs** [comment]
Jenkins, test again

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-dmn/pom.xml [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.kie.server</groupId>
+      <artifactId>kie-server</artifactId>
+      <classifier>ee7</classifier>
+      <type>war</type>
+      <scope>test</scope>
+    </dependency>
+    
+    <!-- DO NOT REMOVE THIS COMMENT. Ref commit 3a92c62 of PR droolsjbpm/droolsjbpm-integration#447
```
Comment (tarilabs) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-dmn/pom.xml]: @etirelli this is finally green passing Jenkins CI; I details my finding for posterity and hopefully shall help others or our future selves for reference.

The error reported:
```
23:46:06 [INFO] KIE :: Execution Server :: Tests ................... SUCCESS [  1.731 s]
23:46:06 [INFO] KIE :: Execution Server :: Tests :: Common Integration Tests SUCCESS [02:48 min]
23:46:06 [INFO] KIE :: Execution Server :: Tests :: Drools Integration Tests SUCCESS [02:59 min]
23:46:06 [INFO] KIE :: Execution Server :: Tests :: jBPM Integration Tests SUCCESS [12:11 min]
23:46:06 [INFO] KIE :: Execution Server :: Tests :: OptaPlanner Integration Tests FAILURE [02:15 min]
23:46:06 [INFO] KIE :: Execution Server :: Tests :: Drools and jBPM Integration Tests SKIPPED
23:46:06 [INFO] KIE :: Execution Server :: Tests :: Controller Integration Tests SKIPPED
23:46:06 [INFO] KIE :: Execution Server :: Tests :: Policies Integration Tests SKIPPED
23:46:06 [INFO] KIE :: Execution Server :: Tests :: Router Integration Tests SKIPPED
23:46:06 [INFO] KIE :: Execution Server :: Tests :: Case Id Generator Integration Tests SKIPPED
23:46:06 [INFO] KIE :: Execution Server :: Tests :: DMN Integration Tests SUCCESS [01:02 min]
23:46:06 [INFO] KIE :: OSGi Integration Parent Project ............. SUCCESS [  0.477 s]
23:46:06 [INFO] kie-karaf-itests-domain-model ...................... SUCCESS [  8.390 s]
23:46:06 [INFO] KIE :: Karaf Integration Tests Kjar ................ SUCCESS [  8.719 s]
23:46:06 [INFO] KIE :: OSGi Integration ............................ SUCCESS [ 10.653 s]
23:46:06 [INFO] KIE :: Karaf Features .............................. SUCCESS [  9.899 s]
23:46:06 [INFO] KIE :: Karaf Integration Tests ..................... SUCCESS [02:54 min]
23:46:06 [INFO] KIE Performance Kit ................................ SUCCESS [ 17.782 s]
23:46:06 [INFO] Drools :: Android Integration ...................... SUCCESS [ 17.497 s]
23:46:06 [INFO] kie-security ....................................... SUCCESS [  8.513 s]
23:46:06 [INFO] Drools and jBPM integration distribution ........... SUCCESS [  7.143 s]
23:46:06 [INFO] BPMS distribution .................................. SUCCESS [ 17.775 s]
23:46:06 [INFO] BRMS distribution .................................. SUCCESS [  9.288 s]
23:46:06 [INFO] ------------------------------------------------------------------------
23:46:06 [INFO] BUILD FAILURE
23:46:06 [INFO] ------------------------------------------------------------------------
23:46:06 [INFO] Total time: 27:52 min (Wall Clock)
23:46:06 [INFO] Finished at: 2017-03-02T18:46:05-05:00
23:46:07 [INFO] Final Memory: 510M/1609M
23:46:07 [INFO] ------------------------------------------------------------------------
23:46:07 [ERROR] Failed to execute goal org.codehaus.cargo:cargo-maven2-plugin:1.6.2:stop (stop-container) on project kie-server-integ-tests-optaplanner: Execution stop-container of goal org.codehaus.cargo:cargo-maven2-plugin:1.6.2:stop failed: Failed to stop the WildFly 10.0.0.Final container. Failure when invoking CLI script, java returned 1 -> [Help 1]
```
had nothing to to with `kie-server-integ-tests-optaplanner` because with this PR I didn't change anything about optaplanner modules.

I then realised the real problem was that `integ-test-xxx` reactor modules need to be executed sequentially, otherwise they would fail for conflicting ports. The way to achieve this is to have each `integ-test-xxx` reactor modules with an "artificial dependency", which is implemented with the lines here.
Ref commit 3a92c62 of PR droolsjbpm/droolsjbpm-integration#447 

In this case, I looked up the actual reactor module sequence, because this `integ-test-dmn` comes after  `kie-server-integ-tests-case-id-generator` in the reactor sequence, the artificial dependency in this dmn module reference case-id-generator, as line 178.

What took me time to figure it out, is that another commit 22f0dc0c may have accidentally removed the comment explaining this "trick" from the one existing pom which I used as a reference during my development, so I'll raise separately to instantiate the comment back (and the reason in this pom I've put the explicit reference possibly helpful for posterity).

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-dmn/src/main/java/org/kie/server/remote/rest/dmn/ModelEvaluatorResource.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                return createCorrectVariant(marshallerHelper, containerId, result, headers, Response.Status.OK, conversationIdHeader );
+            }
+            return createCorrectVariant(marshallerHelper, containerId, result, headers, Response.Status.NOT_FOUND, conversationIdHeader );
+        }  catch (Exception e) {
+            LOG.error("Unexpected error retrieving models. Message: '{}'", e.getMessage(), e);
+            return internalServerError(MessageFormat.format( "ERROR", e.getMessage()), v, conversationIdHeader);
+        }
+    }
+
+    @POST
```
Comment (etirelli) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-dmn/src/main/java/org/kie/server/remote/rest/dmn/ModelEvaluatorResource.java]: @tarilabs I would like to discuss this further next week. I think we might need to segregate the URIs a bit further here.

File: kie-server-parent/kie-server-services/kie-server-services-dmn/src/main/java/org/kie/server/services/dmn/ModelEvaluatorServiceBase.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            
+            DMNContext dmnContext = DMNFactory.newContext();
+            for ( Entry<String, Object> e : evalCtx.getDmnContext().entrySet() ) {
+                dmnContext.set(e.getKey(), e.getValue());
+            }
+            LOG.info("Will use dmnContext: {}", dmnContext);
+            
+            DMNResult result = dmnRuntime.evaluateAll(model, dmnContext);
+            
+            LOG.info("Result:");
```
Comment (etirelli) [inline kie-server-parent/kie-server-services/kie-server-services-dmn/src/main/java/org/kie/server/services/dmn/ModelEvaluatorServiceBase.java]: these messages here should probably be log.debug() instead of info().

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-dmn/src/test/java/org/kie/server/integrationtests/dmn/DMNIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Override
+    protected void addExtraCustomClasses(Map<String, Class<?>> extraClasses)
+            throws Exception {
+
+        // no extra classes.
+    }
+
+    @Test
+    public void testHelloWorld() {
```
Comment (etirelli) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-dmn/src/test/java/org/kie/server/integrationtests/dmn/DMNIntegrationTest.java]: I think we need to create a few more tests covering a few additional scenarios, like error scenarios, null inputs and results, etc.

**Comment by etirelli** [comment]
@tarilabs I think this is a good first step. We will need to further refine this implementation, but we can merge this now and work on additional improvements on a separate PR.

+1
